packages/compiler-cli/src/ngtsc/annotations/src/util.ts

```diff
@@ -375,4 +375,46 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
  */
 export function wrapFunctionExpressionsInParens(expression: ts.Expression): ts.Expression {
   return ts.transform(expression, [parensWrapperTransformerFactory]).transformed[0];
-}
\ No newline at end of file
+}
+
+/**
+ * Resolves the given `rawProviders` into `ClassDeclarations` and returns
+ * a set containing those that are known to require a factory definition.
+ * @param rawProviders Expression that declared the providers array in the source.
+ */
+export function resolveProvidersRequiringFactory(
+    rawProviders: ts.Expression, reflector: ReflectionHost,
+    evaluator: PartialEvaluator): Set<Reference<ClassDeclaration>> {
+  const providers = new Set<Reference<ClassDeclaration>>();
+  const resolvedProviders = evaluator.evaluate(rawProviders);
+
+  if (Array.isArray(resolvedProviders)) {
+    resolvedProviders.forEach(function processProviders(provider) {
+      let tokenClass: Reference|null = null;
+
+      if (Array.isArray(provider)) {
+        // If we ran into an array, recurse into it until we've resolve all the classes.
+        provider.forEach(processProviders);
+      } else if (provider instanceof Reference) {
+        tokenClass = provider;
+      } else if (provider instanceof Map && provider.has('useClass')) {
```

Comment:
The g3 presubmit brought up an interesting case:

```typescript
{provide: Foo, useClass: Bar, deps: [DepA, DepB]}
```

is perfectly fine even if `Bar` is not injectable, as the user is explicitly telling the injector how to call the constructor. So `useClass` with `deps` can be ignored here:

```suggestion
      } else if (provider instanceof Map && provider.has('useClass') && !provider.has('deps')) {
```

Please add a test to that effect :)